rekall-core/rekall/plugins/linux/heap_analysis.py

```diff
@@ -0,0 +1,4388 @@
+
+        # This function is only reliable, if we have the offset to mp_
+        if not self.mp_:
+            # at least we test arena for being swapped
+            if not deactivate_swap_check:
+                self._check_and_report_arena_for_being_swapped(arena)
+
+            return None
+
+        # max arena value can be adjusted at runtime via mallopt func:
+        # see malloc/malloc.c line 4753 and http://man7.org/linux/man-pages/man3/mallopt.3.html
+        # or on startup via env vars (see also link)
+        # if not, this member is 0
+        arena_max = self.mp_.arena_max
+
+        if arena_max == 0:
+            # The maximum number of arenas is calculated with the macro 'NARENAS_FROM_NCORES'
+            # See malloc/arena.c
+            arena_max = self._get_max_number_of_arenas()
+            cores = self._get_number_of_cores()
+
+            # In the case of one core, there can be one more arena than
+            # the result from 'NARENAS_FROM_NCORES'
+            # See function 'arena_get2' in malloc/arena.c
+            if cores == 1:
+                arena_max += 1
+
+        if arena_max == 0:
+            self.session.logging.warn("The result for arena_max has been 0. This shouldn't be "
+                                      "the case and has to be looked into.")
+
+            if not deactivate_swap_check:
+                # as the following for loop will in this case not loop over any arena, we check
+                # the current arena at least for being swapped
+                self._check_and_report_arena_for_being_swapped(arena)
+
+        curr_arena = arena
+        for _ in range(arena_max):
+            swap_check_result = self._check_and_report_arena_for_being_swapped(curr_arena) if \
+                not deactivate_swap_check else None
+
+            if swap_check_result is not True:
+                curr_arena = curr_arena.next
+                if arena == curr_arena:
+                    return True
+
+            else:
+                break
+
+        return False
+
+
+
+    def __init__(self,
+                 glibc_profile=None,
+                 prevent_chunk_preservation=False,
+                 main_arena=None,
+                 malloc_par=None,
+                 **kwargs):
+        super(HeapAnalysis, self).__init__(**kwargs)
+
+        self._libc_profile = None
+        self._libc_offset = None
+        self.arenas = []
+        self.process_as = None
+
+        # all vmas belonging to the current task
+        self.vmas = None
+
+        # only the vmas that we consider to belong to the current task's heap
+        self.heap_vmas = None
+
+        self._size_sz = None
+        self._malloc_alignment = None
+        self._malloc_align_mask = None
+        self._minsize = None
+
+        self.mp_ = None
+        self.mp_offset = malloc_par
+        self.__mmapped_warnings = set()
+        self.__is_statically_linked = False
+        self.__first_chunk_distance = 0
+
+        self.task = None
```

Comment:
this needs to work together with the cc() plugin. We need to get the current task from the session. See https://github.com/google/rekall/blob/master/rekall-core/rekall/plugins/core.py#L1002